### ISSUE TITLE
Port API diff utils to Zinc to make it work again

### ISSUE THREAD

**Issue by smarter** [sbt/zinc#285]
http://www.scala-sbt.org/0.13/docs/Understanding-Recompilation.html#Debugging+an+interface+representation explains how to get API diffs to debug incremental compilation, but that does not seem to work with zinc currently:
1. Enable debug info for scripted tests: https://github.com/sbt/zinc/issues/283
2. Add/edit the `incOptions.properties` file inside a scripted test directory (https://github.com/sbt/zinc/issues/284), for example `zinc/src/sbt-test/source-dependencies/as-seen-from-a/` to contain:
```scala
apiDebug = true
```
3. `wget https://repo1.maven.org/maven2/com/googlecode/java-diff-utils/diffutils/1.3.0/diffutils-1.3.0.jar`
4. Run `sbt -Dsbt.extraClasspath=diffutils-1.3.0.jar`
5. Run `> scripted source-dependencies/as-seen-from-a`
6. Get:
```scala
[error] You have api debugging enabled but DiffUtils library cannot be found on sbt's classpath (out-1)
```
7. Be very sad. Consider all the things you could do in life instead of trying to debug sbt. Go outside.

**Comment by smarter** [comment]
By the way maybe instead of this insane roundabout way of doing diffs you could consider doing something natively in Scala, feel free to steal https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/util/DiffUtil.scala

**Comment by jvican** [comment]
Hey @smarter, thanks for reporting and the way to reproduce it!
  
I would very much prefer that. We could also reuse some of Nicola's work on that front, I remember his coloured diff was super useful.

**Comment by smarter** [comment]
Yes, Nicolas' stuff is in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/util/DiffUtil.scala, we also adapted it recently to do line diffs so that we can use it when comparing a run test actual output with expected output.

**Comment by jvican** [comment]
The goal of this issue is to port [DiffUtil](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/util/DiffUtil.scala) to Zinc, replacing the current diff infrastructure. The current infrastructure is in the `zincCore` project: https://github.com/sbt/zinc/blob/1.0/internal/zinc-core/src/main/scala/sbt/internal/inc/APIDiff.scala.
  
Why is this ticket useful?
  
Zinc has always exposed a way to log the APIs diff that cause recompilation of sources. These diffs are excellent to identify what's going on, whether a behaviour is a bug, or how the actual incremental compilation works.
  
The current infrastructure does not work, failing with the current error reported in this ticket. The goal of the contributors that takes up this issue is to make the diff work, which should be quite straightforward given that most of the util is already implemented in the Dotty compiler, and it even has colors!
  
This is an important ticket to work on, enjoy!

**Comment by jvican** [comment]
This was solved in #346.
